File: C/Partie5/Exo21.c
#include <stdio.h>
#include <stdlib.h>
#include <pthread.h>
#include <unistd.h>
#include <time.h>

#define N 100

double tableau[N];
pthread_mutex_t mutex = PTHREAD_MUTEX_INITIALIZER;

void *thread_moyenne(void *arg)
{
    while (1)
    {
        pthread_mutex_lock(&mutex);
        double copie[N];
        for (int i = 0; i < N; i++)
            copie[i] = tableau[i];
        for (int i = 1; i <= N - 2; i++)
        {
            tableau[i] = (copie[i - 1] + copie[i] + copie[i + 1]) / 3.0;
        }
        pthread_mutex_unlock(&mutex);
        int duree = rand() % 3 + 1;
        sleep(duree);
    }
    return NULL;
}

void *thread_affiche(void *arg)
{
    while (1)
    {
        sleep(4);
        pthread_mutex_lock(&mutex);
        printf("Tableau : ");
        for (int i = 0; i < N; i++)
            printf("%.2f ", tableau[i]);
        printf("\n");
        pthread_mutex_unlock(&mutex);
    }
    return NULL;
}

int main()
{
    srand(time(NULL));
    for (int i = 0; i < N; i++)
    {
        if (i == 0 || i == N - 1)
            tableau[i] = 0.0;
        else
            tableau[i] = (double)(rand() % 101);
    }
    pthread_t t1, t2;
    pthread_create(&t1, NULL, thread_moyenne, NULL);
    pthread_create(&t2, NULL, thread_affiche, NULL);
    pthread_join(t1, NULL);
    pthread_join(t2, NULL);
    return 0;
}

File: Python/Partie5/Exo21.py
import threading
import random
import time

N = 100
tableau = [0.0] * N
mutex = threading.Lock()

def thread_moyenne():
    global tableau
    while True:
        with mutex:
            copie = tableau[:]
            for i in range(1, N-1):
                tableau[i] = (copie[i-1] + copie[i] + copie[i+1]) / 3.0
        time.sleep(random.randint(1, 3))

def thread_affiche():
    while True:
        time.sleep(4)
        with mutex:
            print("Tableau :", [round(x,2) for x in tableau])

# initialisation
for i in range(N):
    if i == 0 or i == N-1:
        tableau[i] = 0.0
    else:
        tableau[i] = random.randint(0, 100)

t1 = threading.Thread(target=thread_moyenne, daemon=True)
t2 = threading.Thread(target=thread_affiche, daemon=True)
t1.start()
t2.start()
# laisser tourner (Ctrl+C pour arrêter)
try:
    while True:
        time.sleep(1)
except KeyboardInterrupt:
    pass